[10.x] Add a new `Numberable` helper

emmadesilva opened the pull request:
## Abstract

Building on https://github.com/laravel/framework/pull/48845, this class will work similarly to the `Stringable` class, but for numbers.

This was requested on the Discord and by several people on Twitter, and mentioned in https://github.com/laravel/framework/pull/48849

### Development

Feedback and contributions are much appreciated! I can be reached here, or in the `#internals` Discord.

- [x] Add accessors for the `Number` class
- [x] Add methods to interact with the value
- [x] Add `Number::of()` and `number()` shorthands
- [ ] See if we can find a better name for this helper? (Numerable has been suggested)

emmadesilva committed:
Create a new Numberable helper

Building on https://github.com/laravel/framework/pull/48845, this class will work similarly to the `Stringable` class, but for numbers.

emmadesilva committed:
Add arithmetic operation shorthands

I'm making so that these helpers modify the underlying value, instead of creating new instances like in the Stringable class. This is because I think modifying the values is more helpful for the kinds of use cases I see for this feature.

emmadesilva committed:
Add accessors for the Number class

emmadesilva committed:
Test trait methods and fluency operations

emmadesilva committed:
Formatting

emmadesilva committed:
Add a `Number::of` helper to create a `Numberable` object

emmadesilva committed:
Add a `number()` helper to create a `Numberable` object

Based on https://github.com/laravel/framework/commit/d41e88519a6e4203b0986a40c5ac8670a157cf59, instead of the newer https://github.com/laravel/framework/pull/40524/commits/a75dea9d059896976f10f8296e00b850faff3c0f as there are no facade methods where the latter functionality adds value to this case.

emmadesilva committed:
Merge remote-tracking branch 'upstream/10.x' into numberable-helper

emmadesilva committed:
Add wrappers for new spell and ordinal helpers

rmunate commented:
https://github.com/laravel/framework/pull/48845#issuecomment-1817535149

maartenpaauw commented:
Would be a great addition, but wouldn't it be better to make this class immutable like the stringable class?

caendesilva commented:
> Would be a great addition, but wouldn't it be better to make this class immutable like the stringable class?

Thanks for the feedback! It definitely could be, but I went this route because I personally think modifying the values is more helpful for the kinds of use cases I see for this feature being used. But I am 100% open to changing it to be immutable if that turns out to be the consensus on what's best.

bert-w commented:
This makes every equation like `($x + 5)/3` unreadable when you write it in functional fluent mode `Numberable($x)->add(5)->divide(3)`.

I would never use this, unless there's actually logic inside that would be worth abstractifying like vector or matrix math.

caendesilva commented:
> This makes every equation like `($x + 5)/3` unreadable when you write it in functional fluent mode `Numberable($x)->add(5)->divide(3)`.
> 
> I would never use this, unless there's actually logic inside that would be worth abstractifying like vector or matrix math.

That's a great point, and a strong reason to use immutable versions. I'll make a patch tomorrow.

Rizky92 commented:
I'm not sure about this PR. `Stringable` works because there is a need for complex string manipulation fluently.

Your previous PR is great for converting unit of measurement or making numeric values readable, but not for manipulating it.

@bert-w's comments is a strong concern. Unless you provide another abstraction for working with arithmetics, I don't really see a good thing why this PR is necessary.

caendesilva commented:
> I'm not sure about this PR. `Stringable` works because there is a need for complex string manipulation fluently.
> 
> Your previous PR is great for converting unit of measurement or making numeric values readable, but not for manipulating it.
> 
> @bert-w's comments is a strong concern. Unless you provide another abstraction for working with arithmetics, I don't really see a good thing why this PR is necessary.

Good points as well! Thanks for all the feedback! Going to try to refine this tomorrow, then open up for review to see what the team thinks. I appreciate your thoughts!

huangdijia commented on tests/Support/SupportNumberableTest.php:
```diff
@@ -0,0 +1,245 @@
+    public function testCurrency()
+    {
+        $this->assertSame($this->numberable(10)->currency(), Number::currency(10));
+        $this->assertSame($this->numberable(1.234)->currency(2), Number::currency(1.234, 2));
```
These two testcases should fail.

caendesilva replied:
My bad. Fixing it now.

gocanto commented:
Sincere and humbre question

Did you evaluate the possibility to use [Money/Money](https://github.com/moneyphp/money/tree/master)? - This lib is battle tested and offer this [functionality](https://github.com/moneyphp/money/blob/master/src/Number.php) out of the box.

cc @taylorotwell @driesvints

emmadesilva committed:
Merge remote-tracking branch 'upstream/10.x' into numberable-helper

emmadesilva committed:
Fix test comparisons

See https://github.com/laravel/framework/pull/49036#discussion_r1398521983

caendesilva commented:
> 🤝 Sincere and humble question,
> 
> Did you evaluate the possibility to use [Money/Money](https://github.com/moneyphp/money/tree/master)? - This lib is battle tested and offer this [functionality](https://github.com/moneyphp/money/blob/master/src/Number.php) out of the box.
> 
> cc @taylorotwell @driesvints

I haven't considered it for this feature, as I think these helpers are best used for quick formatting. If one needs something more robust, I think that's when you would reach for a library like yours. In my opinion, specialised code like that does not need to be in the Laravel core as we have packages to get the functionality when needed.

emmadesilva committed:
Make numberable helpers immutable

Consensus reached in the pull request: https://github.com/laravel/framework/pull/49036

It also removes the tappability as that loses it's functionality when immutable.

caendesilva commented:
> I don't really think this is necessary yet. It's common to chain many string operations together when working with strings - I don't think the same is really the case with numbers.

Fair enough! It was requested by some people after my last PR, and it was fairly quick to make this one so I thought I'd throw it out there in case anyone finds it useful. Thanks for taking the time to review it!
